### PR TITLE
Allow destructuring policies to transform reflection types

### PR DESCRIPTION
**Pull request by adamchester**
See #723 for details.

### CODE REVIEW

**Comment by nblumhardt** [comment]
This looks like the right thing to do :+1:

I need to run my eyes over it again with some more time, will aim to have this turned around tomorrow morning.

File: src/Serilog/ReflectionTypesScalarDestructuringPolicy.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+using Serilog.Core;
+using Serilog.Events;
+
+namespace Serilog.Policies
+{
+    class ReflectionTypesScalarDestructuringPolicy : IDestructuringPolicy
+    {
+        public bool TryDestructure(object value, ILogEventPropertyValueFactory propertyValueFactory, out LogEventPropertyValue result)
+        {
+            var type = value as System.Type;
```
Comment (nblumhardt) [inline src/Serilog/ReflectionTypesScalarDestructuringPolicy.cs]: Any reason not to keep the impl from above? Seems equivalent and a bit more terse:

```csharp
           if (value is Type || value is MemberInfo)
           {
               result = new ScalarValue(value);
               return true;
           }

            result = null;
            return false;
```

Comment (adamchester) [inline src/Serilog/ReflectionTypesScalarDestructuringPolicy.cs]: Good point, will do that. Original is clearer.

**Comment by adamchester** [comment]
Since then, I've thought this PR title is misleading; reflection types are still scalar by default after these changes.

File: src/Serilog/ReflectionTypesScalarDestructuringPolicy.cs (hunk 1 above)
Comment (adamchester) [inline src/Serilog/ReflectionTypesScalarDestructuringPolicy.cs]: Force-pushed in 29c14d0

giphy

File: src/Serilog/Configuration/LoggerDestructuringConfiguration.cs [hunk 2]
```diff
@@ -114,6 +115,24 @@ public LoggerConfiguration ByTransforming<TValue>(Func<TValue, object> transform
...
         /// <summary>
+        /// When destructuring objects, transform instances derived from the specified type with
+        /// the provided function.
+        /// </summary>
+        /// <param name="transformation">Function mapping instances of <typeparamref name="TValue"/>
+        /// to an alternative representation.</param>
+        /// <typeparam name="TValue">The base class Type of values to transform.</typeparam>
+        /// <returns>Configuration object allowing method chaining.</returns>
+        /// <exception cref="ArgumentNullException"></exception>
+        public LoggerConfiguration ByTransformingDerived<TValue>(Func<TValue, object> transformation)
```
Comment (adamchester) [inline src/Serilog/Configuration/LoggerDestructuringConfiguration.cs]: This name might need some more thought

**Comment by nblumhardt** [comment]
I think `ByTransformingDerived` is an interesting addition, but I'd prefer keeping it separate from this PR since the reflection types scenario is unblocked without this change. Works for you? Cheers!

**Comment by adamchester** [comment]
OK - but the issue is that at runtime (in most cases I tested), you never actually get an instance of `System.Type` (you get, for example, an internal `System.RuntimeType`):

image

**Comment by adamchester** [comment]
What that means, of course, is that you can't ever use `Destructure.ByTransforming` to match reflection types in reality.

**Comment by nblumhardt** [comment]
Yep, gotcha :+1: - but with the first PR'd change it's possible to plug in an `IDestructuringPolicy` to cover the scenario, so still a step forward.

`ByTransformingDerived` is probably an improvement in itself, but looking at other places it'd be used might suggest some API tweaks or alternatives we won't spot just looking at reflection types. I'm also (guiltily) interested in the possibility of `ByTransforming(..., includeSubtypes: true)` etc. Trying to keep all of that discussion from derailing the PR here :-)

**Comment by adamchester** [comment]
Ahh! Didn't remember `IDestructuringPolicy` was public :+1: 

Will remove `ByTransformingDerived` because it's possible to use without it.

**Comment by adamchester** [comment]
Removed all trace of `ByTransformingDerived` now.

**Comment by nblumhardt** [comment]
Awesome, thanks!

**Comment by adamchester** [comment]
image

**Pull request merged**